src/error/mod.rs

```diff
@@ -0,0 +1,149 @@
+// This file is part of Twig (ported to Rust).
+//
+// For the copyright and license information, please view the LICENSE
+// file that was distributed with this source code.
+
+/// Twig base exception
+
+use std::fmt::{self, Display};
+use std::convert::Into;
+
+
+#[macro_use]
+pub mod macros;
+// use std Error-trait to improve cross-crate compatibility
+// don't mix it up with Err(X)
+pub use std::error::Error as StdError;
+pub mod api;
+
+// generic wrapper around some StdError - adds location support
+#[derive(Debug)]
+pub struct Error<T>
+    where T: api::ErrorCode
+{
+    // the exception codes are going to be enums
+    // - i.e. Exception<MY_ENUM> implements std::error::Error without any boilerplate
+    // to MY_ENUM. Hurray! :-)
+    code: T,
+    // I decided to call this field `code` instead of `error` to not confuse it with the Error trait
+    location: Location,
+    // chaining is required by std::error::Error
+    cause: Option<Box<StdError>>,
```

Comment:
Idiomatic way is to include `std::error` as `self` and use `error::Error` instead of `StdError`.